### DISCUSSION TITLE
Keyed collection, using keyBy, loses keys on submit

### DISCUSSION THREAD

**Discussion by notflip** [livewire/livewire#3142] [7. Bugs (archived)] [answered]
### Description

When I set a variable to a collection that uses keyBy, on a submit the keys are lost.

### Exact steps to reproduce

- in mount() set a variable to a keyed collection using `$this->allProducts = Product::all()->keyBy('id');`
- add a submit handler to a form using `<form wire:submit.prevent="submitForm">`
- dd() $this->allProducts in the submitForm() method
- it returns a collection without the keys

### Stripped-down, copy-pastable code snippets

```
public function mount()
{
    $this->allProducts = Product::all()->keyBy('id');
}

public function submit()
{
    dd($this->allProducts);
}

```

**Context**
- Livewire version: 1.3.5
- Laravel version: 6.18
- Browser: Irrelevant

**Comment by Jamesking56** [comment]
Just got this without the submit part, simply referring to the collection in the view shows the keys missing as well. I think it must be to do with how Collections are serialised in component data.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by Jamesking56** [comment]
REOPEN

I can probably build at least a failing test for this, I'll add it to my list

**Comment by Jamesking56** [comment]
REOPEN

**Comment by glaesser** [comment]
> Just got this without the submit part, simply referring to the collection in the view shows the keys missing as well. I think it must be to do with how Collections are serialised in component data.

Just ran into this as well...seems like it's not possible to to use keyBy on a collection at all, as the keys are stripped in the view?

**Comment by glaesser** [comment]
@calebporzio would it be possible to address this in the next release?

FYI, this issue applies to latest Livewire version (2.3.1) and Laravel 8.

**Comment by austenc** [comment]
@notflip @Jamesking56 @glaesser 

Is this still happening for you folks on the latest version of Livewire (`2.3.6`)?

**Comment by xtfer** [comment]
@austenc Still happening with Livewire 2.10.1

**Comment by joshhanley** [comment, marked as answer]
This issue with this is Livewire makes use of Laravel's serialiser that it uses for queues. As such it doesn't have support for custom things like `keyBy`. All it keeps track of is the model name and the IDs of the models in the collection.

If you need `keyBy` I would make use of a combination of a property that stores the unkeyed collection and a computed property which takes the collection and keys it.

For example I'll use products as per the original comment. I've renamed the `allProducts` property to `products`, allowing us to use `allProducts` for the computed property.

```php
public $products

public function mount()
{
    $this->products = Product::all();
}

public function getAllProductsProperty()
{
    return $this->products->keyBy('id');
}

public function submit()
{
    dd($this->allProducts);
}
```

Hope this helps!

**Reply by xtfer** [reply]
Thanks @joshhanley. That makes sense and is a good solution.

**Reply by joshhanley** [reply]
No worries! 🙂
